src/dataset.jl

```diff
@@ -1013,9 +1013,20 @@ function buildoverviews!(
     return dataset
 end
 
-function destroy(dataset::AbstractDataset)::Nothing
-    GDAL.gdalclose(dataset)
+# TODO: Wrap `GDAL.CPLErr`
+function close(dataset::AbstractDataset)::GDAL.CPLErr
```

Comment:
In the case of OGR, a dataset might have features (from e.g. https://github.com/yeesian/ArchGDAL.jl/blob/4b40fe857503ec2ef8bc431919d8d31bcec7cba5/src/ogr/featurelayer.jl#L29-L41). There we took the approach of tracking references using an attribute named `.ownedby` (so that Julia's GC wouldn't finalize the dataset while there are references to it) and have an implicit convention that users shouldn't be calling `destroy` themselves (and either rely on Julia's GC, or the do-block approach for managing context).

That said, if you feel strongly about it, I'm sympathetic to the argument for being able to "force close" a dataset that resulted in the `.children` attribute being introduced. To avoid https://gdal.org/api/python_gotchas.html#python-crashes-or-throws-an-exception-if-you-use-an-object-after-deleting-a-related-object, should we introduce a new type for datasets that tracks their children, and restrict this function to only accept those datasets?

Reply:
The issue here is this https://github.com/OSGeo/gdal/pull/10490 . When writing a dataset, there is no reliable way to ensure it has been written since one has to wait until all GDAL objects pointing into the dataset have been garbage collected and finalized.

One way out would be not provide an interactive API, and to handle all lifetimes manually or via context managers. That's quite inconvenient in many cases.

Thus I decided to implement a `close` function that (by default) hard-closes a dataset when the dataset is writable. For this, the dataset needs to hold references that need to be released when the dataset is force-closed.

This has nothing to do with lifetime management. You can open datasets with a `hard_close=false` option, and no such tracking happens. However, it's then unclear when a dataset will actually be written to disk.

I now think that the dataset should hold weak references (not strong ones) to the objects that should be released. I'll update the code.

If you have a different suggestion for reliably closing or flushing a dataset I'd be happy to change things.

Reply:
> One way out would be not provide an interactive API, and to handle all lifetimes manually or via context managers. That's quite inconvenient in many cases.

Yeah, it is inconvenient. I'd still recommend it for this PR though -- to keep the PR scoped to the introduction of the MDArray API and to introduce the hard close option in a separate PR.

> For this, the dataset needs to hold references that need to be released when the dataset is force-closed.

Even if that were implemented correctly, it still doesn't address the resulting complexity of making all GDAL objects suspect of potentially corresponding to resources that might have already been released.

Reply:
I will have to think about how to proceed without creating very unexpected behaviour.

There is already a function `isnull` that checks whether a reference is valid. References become invalid when they are released.